src/backend/distributed/master/master_stage_protocol.c

```diff
@@ -427,31 +433,58 @@ CreateShardPlacements(Oid relationId, int64 shardId, List *ddlEventList,
 		ereport(ERROR, (errmsg("could only create %u of %u of required shard replicas",
 							   placementsCreated, replicationFactor)));
 	}
+
+	return connectionList;
 }
 
 
 /*
  * WorkerCreateShard applies DDL commands for the given shardId to create the
- * shard on the worker node. Note that this function opens a new connection for
- * each DDL command, and could leave the shard in an half-initialized state.
+ * shard on the worker node. Note that this function opens a new placement
+ * connection if the command is run inside the transaction or the local table
+ * has a data on it. Function does NOT work transactionally for append
+ * distributed tables.
  */
-bool
+MultiConnection *
 WorkerCreateShard(Oid relationId, char *nodeName, uint32 nodePort,
 				  int shardIndex, uint64 shardId, char *newShardOwner,
-				  List *ddlCommandList, List *foreignConstraintCommandList)
+				  List *ddlCommandList, List *foreignConstraintCommandList,
```

Comment:
This function is bulky with 11 parameters and contains a lot of caller-dependent logic. What if we just pass it a `MultiConnection *connection` and do the connection handling outside of this function?